### ISSUE TITLE
Freeze when deleting too many lines in a highlighted file when indent-blankline is installed

### ISSUE THREAD

**Issue by Awpteamoose** [neovide/neovide#2217]
**Describe the bug**
Selecting and deleting too many lines when indent-blankline is installed freezes neovide in an odd way. Pressing keys or the "X" on the window is logged, but nothing happens, resizing the windows causes some shifts in layout but not of the right sort.

**To Reproduce**
1. Install & configure [indent-blankline](https://github.com/lukas-reineke/indent-blankline.nvim)
2. Open a new buffer in neovide
3. Paste a bunch of C code, like 1500 lines, I used https://raw.githubusercontent.com/qpakzk/linux-perf/e9f0dd7123612dbf3f5e4f48904117d4eaf27294/fs/dax.c
4. `:set filetype=c`
5. Visual select about 500 lines and delete them

**Expected behavior**
Selected text is deleted.

**Screenshots**
[vidya instead](https://drive.google.com/file/d/1pHqAi4P37jNZj3jbvUcwNuciXiCfA84y/view?usp=drivesdk)

**Desktop (please complete the following information):**
 - OS: Windows 10
 - Neovide Version: 0.12
 - Neovim Version: 0.9.4 (it's 0.9.2 in the vidya but I updated and it was exactly the same)

**Please run `neovide --log` and paste the contents of the `.log` file created in the current directory here:**
[neovide12_rCURRENT.log](https://github.com/neovide/neovide/files/13788267/neovide12_rCURRENT.log)

**Additional context**
```
[16:13]fredizzimo: Can you try disabling indent-blankline? It keeps redefining the same highlights. And maybe causing a buffer overflow somewhere, the last message we get from Neovim seems to be partial
[16:13]Awpteamoose: can do
[16:15]Awpteamoose: that does fix it
[16:15]Awpteamoose: should I go complain to indent-blankline people?
[16:16]fredizzimo: I think it does something wrong, but there’s probably also another bug somewhere, either in Neovim or Neovide
-- snip --
[16:35]fredizzimo: After looking a bit closer at the log, there’s probably a buffer overflow in Neovim. Since it stops sending and the UI flush message is missing. And before that it sends many batches, trying to split the overelong message.
```

**Comment by fredizzimo** [comment]
Can you test this with v3 of indent-blankline? I believe you are still using v2, because of the name of the highlight groups.

We probably need a log from v3 in order to report a bug.

**Comment by Awpteamoose** [comment]
I did not notice that I had v2.2 pinned in my vimrc, it doesn't reproduce with v3 onwards.

**Comment by Awpteamoose** [comment]
I didn't notice that I had v2 pinned in my config as v3 doesn't work the way I want it to. I switched to https://github.com/shellRaining/hlchunk.nvim and the issue went away.

**Comment by alxndr13** [comment]
had a similar issue, inserted about 25000 lines. hangs when: `:g/debug/d`

removing indent-blankline fixes the issue. was using `v3.3.7`

**Comment by yavorski** [comment]
@echasnovski just FYI.
This issue is also valid for "[mini.indentscope](https://github.com/echasnovski/mini.indentscope)".  
Visual select + joining lines is freezing the neovide editor too when mini.indentscope is enabled.

**Comment by echasnovski** [comment]
> @echasnovski just FYI. This issue is also valid for "[mini.indentscope](https://github.com/echasnovski/mini.indentscope)". Visual select + joining lines is freezing the neovide editor too when mini.indentscope is enabled.

Does this happen in plain TUI? And how many lines are we talking here?

**Comment by fredizzimo** [comment]
@echasnovski The bug seems to be unique to ui clients that enable the ui extension `ext_hlstate` https://neovim.io/doc/user/ui.html#ui-hlstate

I think the actual bug is in Neovim, but something in the plugins trigger it. The only other information I have can be seen in the original log, it keeps sending `hl_attr_define` for the same highlight group, until the message apparently gets too big and neovim stops sending the rest. It's already split into multiple messages, but that does not seem to be enough, and there's no `flush`

Unfortunately, I wont have a development environment available for debugging this further until the end of the month, so this is probably all info I can give until then.

**Comment by echasnovski** [comment]
I see.

What 'mini.indentscope' (and 'indent-blankline', at least before v3) is doing is adding and removing quite a lot of extmarks. Groups are defined just once.

Initially I thought it might be an issue with computation of scope (which might be slow-ish, but for me starting from around 100000), but that mostly introducing lag. If it is not reproducible in plain TUI, then i think it indeed should be an issue with Neovim <-> Neovide communication.

**Comment by yavorski** [comment]
TUI works fine. To reproduce the issue, here is an [example SVG file](https://dev.w3.org/SVG/tools/svgweb/samples/svg-files/AJ_Digital_Camera.svg). Simply paste its contents into the Neovide buffer, then `V`isually select the `defs` tag spanning lines 50 to 257 and press `J`. Neovide will experience a freeze. The issue is reproducible on Linux as well.

**Comment by fredizzimo** [comment]
Could someone create a minimal reproduktion template with `mini.indentscope`? 
https://github.com/neovim/neovim/blob/master/contrib/minimal.lua

We also need a logfile with that.

Then we should be able to properly report the bug to Neovim.

It might even be possible to use  https://neovim.io/doc/user/lua.html#vim.ui_attach(), to make it independent from Neovide. Note that ext_hlstate needs to be enabled

**Comment by yavorski** [comment]
## `minimal.lua`

```lua
-- Run this file as `nvim --clean -u minimal.lua`
-- Run this file as `neovide --log -- --clean -u minimal.lua test.svg`

for name, url in pairs{
  -- ADD PLUGINS _NECESSARY_ TO REPRODUCE THE ISSUE, e.g:
  -- some_plugin = 'https://github.com/author/plugin.nvim'
  mini_indentscope = 'https://github.com/echasnovski/mini.indentscope'
} do
  local install_path = vim.fn.fnamemodify('nvim_issue/'..name, ':p')
  if vim.fn.isdirectory(install_path) == 0 then
    vim.fn.system { 'git', 'clone', '--depth=1', url, install_path }
  end
  vim.opt.runtimepath:append(install_path)
end

-- ADD INIT.LUA SETTINGS _NECESSARY_ FOR REPRODUCING THE ISSUE
require('mini.indentscope').setup({})
```

#### LOG FILE (RAW)

* [https://pastebin.pl/view/raw/f9f1428f](https://pastebin.pl/view/raw/f9f1428f)
* [SVG File](https://dev.w3.org/SVG/tools/svgweb/samples/svg-files/AJ_Digital_Camera.svg)

Not sure how to do rest with the `ext_hlstate` ... can't find any simple example, if you point out any it would be great.

**Comment by mawkler** [comment]
I have the same issue 🙂

**Comment by fredizzimo** [comment]
I reported this issue
* https://github.com/neovim/neovim/issues/26980

**Comment by shivamchandra75** [comment]
> I didn't notice that I had v2 pinned in my config as v3 doesn't work the way I want it to. I switched to https://github.com/shellRaining/hlchunk.nvim and the issue went away.

@Awpteamoose 
can you please share your config for `hlchunk.nvim`, for me the plugin also gives the same freezing behaviour with default setup. when i press di{ or any operator pending command fast enough this hangs neovide.

**Comment by pluiedev** [comment]
Can repro with deleting as few as 150-ish lines of JavaScript... oh well, just gonna delete things piece by piece for now given that it's an upstream issue

**Comment by fredizzimo** [comment]
The fix is now included in Neovim master and should be included in the next nightly build.

I won't mark this issue as fixed though, since it will still freeze for a short while, when deleting the lines, but it should no longer totally hang.

**Comment by smithwinston** [comment]
I just ran into this myself deleting 53 lines from a C program (Neovide 0.12.2 and nvim 0.9.5).

**Comment by fredizzimo** [comment]
Neovim 0.10.0 is out with the fix included. Furthermore, we released Neovide 0.13.0, which no longer depends on the `ext_hlstate` extension, which is the main culprit for emitting a lot of additional information.
